src/hotspot/share/oops/fieldInfo.cpp

```diff
@@ -113,30 +119,161 @@ Array<u1>* FieldInfoStream::create_FieldInfoStream(GrowableArray<FieldInfo>* fie
+    Symbol *name;
+    Symbol *signature;
+    int index;
+    int position;
+  } field_pos_t;
+  field_pos_t *positions = nullptr;
+
+  positions = NEW_RESOURCE_ARRAY(field_pos_t, java_fields);
+  for (int i = 0; i < java_fields; ++i) {
+    assert(r.has_next(), "number of fields must match");
+
+    positions[i].position = r.position();
+    FieldInfo fi;
+    r.read_field_info(fi);
+
+    positions[i].name = fi.name(cp);
+    positions[i].signature = fi.signature(cp);
+    positions[i].index = i;
+  }
+  auto compare_pair = [](const void *v1, const void *v2) {
+    const field_pos_t *p1 = reinterpret_cast<const field_pos_t *>(v1);
+    const field_pos_t *p2 = reinterpret_cast<const field_pos_t *>(v2);
+    return compare_name_and_sig(p1->name, p1->signature, p2->name, p2->signature);
+  };
+  qsort(positions, java_fields, sizeof(field_pos_t), compare_pair);
+
+  auto write_position = position_width == 2 ?
+    [](u1 *ptr, int position) { *reinterpret_cast<u2*>(ptr) = checked_cast<u2>(position); } :
+    [](u1 *ptr, int position) {
+      ptr[0] = static_cast<u1>(position);
+      ptr[1] = static_cast<u1>(position >> 8);
+      ptr[2] = checked_cast<u1>(position >> 16);
+    };
+  auto write_index = index_width == 1 ?
+    [](u1 *ptr, int index) { *ptr = checked_cast<u1>(index); } :
+    [](u1 *ptr, int index) { *reinterpret_cast<u2 *>(ptr) = checked_cast<u2>(index); };
+  for (int i = 0; i < java_fields; ++i) {
+    u1 *ptr = table->adr_at(item_width * i);
+    write_position(ptr, positions[i].position);
+    write_index(ptr + position_width, positions[i].index);
+  }
+  return table;
+}
+
 GrowableArray<FieldInfo>* FieldInfoStream::create_FieldInfoArray(const Array<u1>* fis, int* java_fields_count, int* injected_fields_count) {
-  int length = FieldInfoStream::num_total_fields(fis);
-  GrowableArray<FieldInfo>* array = new GrowableArray<FieldInfo>(length);
   FieldInfoReader r(fis);
-  *java_fields_count = r.next_uint();
-  *injected_fields_count = r.next_uint();
+  r.read_field_counts(java_fields_count, injected_fields_count);
+  int length = *java_fields_count + *injected_fields_count;
+
+  GrowableArray<FieldInfo>* array = new GrowableArray<FieldInfo>(length);
   while (r.has_next()) {
     FieldInfo fi;
     r.read_field_info(fi);
     array->append(fi);
   }
   assert(array->length() == length, "Must be");
-  assert(array->length() == *java_fields_count + *injected_fields_count, "Must be");
   return array;
 }
 
 void FieldInfoStream::print_from_fieldinfo_stream(Array<u1>* fis, outputStream* os, ConstantPool* cp) {
-  int length = FieldInfoStream::num_total_fields(fis);
   FieldInfoReader r(fis);
-  int java_field_count = r.next_uint();
-  int injected_fields_count = r.next_uint();
+  int java_fields_count;
+  int injected_fields_count;
+  r.read_field_counts(&java_fields_count, &injected_fields_count);
   while (r.has_next()) {
     FieldInfo fi;
     r.read_field_info(fi);
     fi.print(os, cp);
   }
 }
+
+#ifdef ASSERT
```

Comment:
Thank you for adding this.

Suggest an additional check to validate the binary search:

For each name/sig pair in the field stream, call the lookup function (or its subroutine) and ensure that it steers to the right position in the field stream.

Reply:
Done. This was really useful; I found a stupid error that caused reading in bad index if that was >= 256.